### DISCUSSION TITLE
Include a CLI with app

### DISCUSSION THREAD

**Discussion by manterfield** [wailsapp/wails#3098] [Q&A] [answered]
This might be a dumb question, so apologies if so. I couldn't find anything in docs and I'm not yet 100% comfortable with how Wails builds/deploys to figure it out from scratch.

We're looking at moving our browser/CLI based app to Wails and distributing as a desktop app, but we'd still want a CLI interface too. Ideally users wouldn't have to download two separate things.

Some apps I have installed on Mac also have a CLI. Docker desktop is an example.

Is this something we can do today with Wails? If not, I would be grateful for any hacky workarounds people have used to achieve the same result.

**Comment by manterfield** [comment]
Okay, I've seen this project using Wails with a CLI app: https://github.com/raguay/BulletinBoard/tree/main

Though to make that work the user has to add `alias bb="/Applications/BulletinBoard.app/Contents/macOS/BulletinBoard"`

That's _okay_ but if there's a way to avoid that alias it would be amazing. I could live with custom install code adding that or an equivalent setup, but can't see a way to run custom code on install.

**Comment by leaanthony** [comment]
You want your app to work 2 different ways but only download 1 binary? You don't need to have the whole .app directory structure but your app will be limited without it.

**Reply by manterfield** [reply]
Two different interfaces to call the same code, yes. I believe there's a fair few apps that have that capability but I think they do some hackery to make that work.

The primary use case is a desktop app, so the CLI part would bend to those restrictions rather than the other way round. I wouldn't want to muck around with the installed location or anything.

The above alias is acceptable if there's no better way to do things, I just figured it would be something others have done and had some nice ideas on.

Docker for Mac is a great example of this, though they do seem to run users through an auto-install of the CLI tools.

---

My ideal solution gives the following:

- No need to download an extra CLI (it's identical code barring the pieces to serve the UI, so ideally avoided)
- No need for the user to create manual aliases
- App is completely unaffected and still works as Wails is designed to work

---

I suppose on Mac and Linux I could just take the approach from BulletinBoard, but on first run take the user through setup and create a symlink in e.g. /usr/local/bin to the app so it's on the path. That should tick all of the above boxes I think.

I've not tried it yet, and I don't know if there's an equivalent I can do for Windows.

**Reply by marcus-crane** [reply, marked as answer]
If you'd like another point of reference, I just went through a bunch of this with my own Wails app: https://github.com/marcus-crane/october/releases/tag/v1.9.0-beta1

Figuring out codesigning for macOS was a bit of a pain since the CLI needs to be signed before being inserted into the `.app` for notarisation if you're doing that whole process.

For Windows, I slightly modify the default `nsi` project to include the CLI in the installation directory while for Linux, I distribute as a `.deb` file which puts the CLI in `/usr/local/bin` which is already (probably) in the user's PATH by default

Personally, as a user, I would just add the location to my PATH (and I have done that for some tools like https://postgresapp.com) rather than doing a whole alias but I agree it feels a bit awkward.

Another route for macOS might be packaging up your CLI as a `.pkg` which installs in `/usr/local/bin`. You could then put that into a dmg where the user can optionally in the CLI and/or drag the `.app` into their Applications folder depending on their preferences.

Presumably for Windows, you could add the ability to modify the user's PATH which is something I've seen various installers offer over the years.

**Reply by manterfield** [reply]
@marcus-crane Really appreciate this, thank you! The extra reference is handy.

We've ended up packaging the CLI with the app in a dmg, though from memory I don't think we've had it install into `/usr/local/bin`, so I'll double check that part. That'll be much neater than mucking about in runtime code to get it on the path.

**Reply by airtonix** [reply]
just use go-releaser. it does all the correct things with regards to putting things in the right place on all OS
